### PR TITLE
div_kernel: throw when dividing by integer zero

### LINKED ISSUE #327: Segmentation fault when dividing by zero with integer tensors
This seems to be an issue with TH and not pytorch, the same thing happens in lua torch.

```python
import torch
a = torch.IntTensor([0,1])
b = torch.IntTensor([0,1])
print(a.div(b)) # Floating point exception (core dumped)
```

Another unrelated weirdness happens with float tensors when the answer is `inf`

```python
import torch
a = torch.FloatTensor([1,1])
b = torch.FloatTensor([0,0])
a.div(b)
```
which works on lua torch, but gives
```
<repr(<torch.FloatTensor at 0x7f8a012eee18>) failed: RuntimeError: inconsistent tensor size at /tmp/pip-zh0a53-build/torch/lib/TH/generic/THTensorMath.c:90>
```
in pytorch.

cc @ezyang @gchanan @zou3519

### PR DESCRIPTION
**Pull request by Baranowski**
Fixes #327

### CODE REVIEW

**Comment by Baranowski** [comment]
Much to my surprise, the additional check does not seem to cause noticeable slowdown. To be fair, the measurements are not stable in my dev environment between different runs (and thus I'm not posting them) but both timeit and ASV suggest that there is nothing outrageous happening (i.e., if there is a slowdown, it's no greater than about 5%).

For the record, here is my ASV benchmark:
```python
class TimeDiv:
    params = [100, 500, 1000, 2000, 3000]

    def setup(self, n):
        self.a = torch.rand(n, n).mul(10000).ceil().int()
        self.b = torch.rand(n, n).mul(10000).ceil().int().add(1)

    def time_div(self, n):
        self.a.div(self.b)
```

And timeit:
```python
a = None
b = None

if __name__ == "__main__":
    for size in (100, 500, 1000, 2000, 3000):
        a = torch.rand(size, size).mul(10000).ceil().int()
        b = torch.rand(size, size).mul(10000).ceil().int().add(1)
        a.div(b)
        print("{size}x{size}: {duration}".format(
            size=size,
            duration=timeit.timeit("a.div(b)", setup="from __main__ import a, b", number=10000)
        ))
```

As for the error handling, division by zero results in the following:
```
$ python div_by_zero.py 
Traceback (most recent call last):
  File "div_by_zero.py", line 4, in <module>
    a.div(a)
RuntimeError: ZeroDivisionError

$ cat div_by_zero.py 
import torch

a = torch.IntTensor([1, 0])
a.div(a)
```

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit 903cdf97:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=903cdf97f7a1e3d4fff51017d5638f8d547d1785).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4343899) pytorch_linux_xenial_py3_6_gcc5_4_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=91057618) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=91057618))
<details>
<summary>
<code>Automatic merge failed; fix conflicts and then commit the result.</code>
</summary>

```
CONFLICT (add/add): Merge conflict in .jenkins/pytorch/win-test-helpers/test_python_all_except_nn.bat 
Auto-merging .jenkins/pytorch/win-test-helpers/test_python_all_except_nn.bat 
CONFLICT (add/add): Merge conflict in .jenkins/pytorch/test.sh 
Auto-merging .jenkins/pytorch/test.sh 
CONFLICT (add/add): Merge conflict in .jenkins/pytorch/multigpu-test.sh 
Auto-merging .jenkins/pytorch/multigpu-test.sh 
CONFLICT (add/add): Merge conflict in .circleci/verbatim-sources/workflows-binary-builds-smoke-subset.yml 
Auto-merging .circleci/verbatim-sources/workflows-binary-builds-smoke-subset.yml 
CONFLICT (add/add): Merge conflict in .circleci/config.yml 
Auto-merging .circleci/config.yml 
Automatic merge failed; fix conflicts and then commit the result. 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

**Review by andreaskoepf — APPROVED** [review]
Looks good to me (see my comment for test_torch.py).

File: test/test_torch.py [hunk 1]
```diff
@@ -14256,6 +14256,13 @@ def test_cat_scalars(self, device):
         with self.assertRaisesRegex(RuntimeError, 'zero-dimensional.*cannot be concatenated'):
             torch.cat([x, y])
 
+    @onlyCPU
+    def test_div_zero(self, device):
```
[the remark this answers is not in the document]
Comment (ezyang) [reply test/test_torch.py]: Yes, there seems to be a bit of a misunderstanding about what the parameters are supposed to do. At the very least you should do the allocations with the device in question (stop using IntTensor!)

**Review by ezyang — CHANGES_REQUESTED** [review]
Need some testing changes.

**Comment by ezyang** [comment]
@VitalyFedyunin or @ngimel could you help us check if there are any perf implications; if the benchmarking above seems sufficient?

**Comment by VitalyFedyunin** [comment]
As per https://www.agner.org/optimize/instruction_tables.pdf conditional short jump is about 20-40 times faster than division op. So slow down seems to be insignificant and benchmarks are correct.

**Comment by VitalyFedyunin** [comment]
I think we can give it go, but prefer to have second opinion CC @ngimel

**Review by VitalyFedyunin — APPROVED** [review]

**Comment by Baranowski** [comment]
As for the CI failure, I cannot reproduce it and it doesn't seem related to my change, so I'm tempted to claim it spurious.

**Review by ezyang — APPROVED** [review]

**Review by ngimel — APPROVED** [review]
lgtm

**Pull request closed without merging**